### PR TITLE
Debug info preservation in loop-unroll pass

### PR DESCRIPTION
**Pull request by jaebaek**
When we copy the loop body to unroll it, we have to copy its
instructions but DebugDeclare or DebugValue used for the declaration
i.e., DebugValue with Deref must not be copied and only the first block
can contain those instructions.

### CODE REVIEW

**Comment by dnovillo** [comment]
> When we copy the loop body to unroll it, we have to copy its
> instructions but DebugDeclare or DebugValue used for the declaration
> i.e., DebugValue with Deref must not be copied and only the first block
> can contain those instructions.

Can you add an example?  What is a `DebugValue` with `Deref`?  And why can't it be copied?

Sorry for all the questions!  I'm still trying to wrap my head around some of these changes.

File: source/opt/loop_unroller.cpp [hunk 1]
```diff
@@ -286,6 +286,10 @@ class LoopUnrollerUtilsImpl {
   // to be the actual value of the phi at that point.
   void LinkLastPhisToStart(Loop* loop) const;
 
+  // Kill DebugDeclare or DebugValue that is used for the declaration with
+  // Deref from |bb|.
```
Comment (dnovillo) [inline source/opt/loop_unroller.cpp:290]: What does 'DebugValue that is used for the declaration' mean?  May be: "Kill all debug declaration instructions from |bb|".  It seems that there are two ways to tie debug info to a variable:

* `DebugDeclare`
* `DebugValue` with a dereference expression.  Used when a compound type is scalarized (e.g. by SROA).

We seem to be overloading terminology.  Perhaps call these something like "debug info declaration" instructions?  It uses the word "declare" again, but I can't think of anything better atm.  It should convey that there are more than one way to associate debug info with a memory location or part of a memory location.  Names are hard ;)

File: source/opt/debug_info_manager.h [hunk 2]
```diff
@@ -148,6 +148,10 @@ class DebugInfoManager {
   // Erases |instr| from data structures of this class.
   void ClearDebugInfo(Instruction* instr);
 
+  // Returns true if |instr| is a DebugDeclare or a DebugValue with Deref
+  // that is used for a declaration.
+  bool IsUsedForDebugDeclare(Instruction* instr);
```
Comment (dnovillo) [inline source/opt/debug_info_manager.h:153]: I don't follow.  Does this mean that the instruction acts as a declaration point?  Used for `DebugDeclare` means to me a `DebugDeclare`.  Perhaps `IsDebugInfoDeclaration`?  See my comment below.

File: source/opt/loop_unroller.cpp (hunk 1 above)
Comment (jaebaek) [reply source/opt/loop_unroller.cpp:290]: "Kill all debug declaration instructions from |bb|" sounds good. Thanks!

File: source/opt/debug_info_manager.h (hunk 2 above)
Comment (jaebaek) [reply source/opt/debug_info_manager.h]: I just changed it to `IsDebugDeclare` that returns true if `instr` is a debug declaration instruction.

**Pull request merged**